fix GetZoneKey nil pointer and add some testcase

kitianFresh opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:
fix nil pointer error in GetZoneKey, or it will lead to scheduler panic when the node is nil
**Which issue(s) this PR fixes**:
fix the nil pointer panic when the param node is nil, and avoid scheduler panic

we find it in our production env, the scheduler panic and restart all the time
![image](https://user-images.githubusercontent.com/8846958/99664644-753e8700-2aa3-11eb-83be-65970e287b69.png)


**Special notes for your reviewer**:
no

**Does this PR introduce a user-facing change?**:
no

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

kitianFresh commented:
/lgtm

kitianFresh committed:
fix conflict

thockin commented:
Do we understand why this is being called with a nil node?

@ahg-g  ?

ahg-g commented:
The scheduler function appearing in the screenshot that invoked GetZoneKey was refactored in v1.18, so I am guessing you are getting this on a v1.17 cluster or lower?

> Do we understand why this is being called with a nil node?

A deleted node: we wrap v1.Node object in a scheduler.NodeInfo type, when a node is deleted, the NodeInfo object may continue to exist until all pods on the node are actually deleted. In 1.18+ we fixed the scheduling cycle such that only NodeInfos with an actual Node is passed through the plugins.

thockin commented:
So we should NOT merge this?  It's probably harmless, but it does imply
that nil is acceptable...

On Fri, Nov 20, 2020 at 3:43 AM ahg-g <notifications@github.com> wrote:

> The scheduler function appearing in the screenshot that invoked GetZoneKey
> was refactored in v1.18, so I am guessing you are getting this on a v1.17
> cluster or lower?
>
> Do we understand why this is being called with a nil node?
>
> A deleted node: we wrap v1.Node object in a scheduler.NodeInfo type, when
> a node is deleted, the NodeInfo object may continue to exist until all pods
> on the node are actually deleted. In 1.18+ we fixed the scheduling cycle
> such that only NodeInfos with an actual Node is passed through the plugins.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/96711#issuecomment-731121125>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVE3H2XGW6UUEUQFY3DSQZI6DANCNFSM4T3IXQBA>
> .
>

ahg-g commented:
If we want to do it, then we should probably do it for all functions in the file. I think we shouldn't merge this.

thockin commented:
@kitianFresh Re-open if you disagree :)

kitianFresh commented:
> If we want to do it, then we should probably do it for all functions in the file. I think we shouldn't merge this.



> The scheduler function appearing in the screenshot that invoked GetZoneKey was refactored in v1.18, so I am guessing you are getting this on a v1.17 cluster or lower?
> 
> > Do we understand why this is being called with a nil node?
> 
> A deleted node: we wrap v1.Node object in a scheduler.NodeInfo type, when a node is deleted, the NodeInfo object may continue to exist until all pods on the node are actually deleted. In 1.18+ we fixed the scheduling cycle such that only NodeInfos with an actual Node is passed through the plugins.



> The scheduler function appearing in the screenshot that invoked GetZoneKey was refactored in v1.18, so I am guessing you are getting this on a v1.17 cluster or lower?
> 
> > Do we understand why this is being called with a nil node?
> 
> A deleted node: we wrap v1.Node object in a scheduler.NodeInfo type, when a node is deleted, the NodeInfo object may continue to exist until all pods on the node are actually deleted. In 1.18+ we fixed the scheduling cycle such that only NodeInfos with an actual Node is passed through the plugins.

yes, we use v1.11, and this bug is because the selector_spreading priority map phase return a HostPriority with empty name, maybe the node is nil, then in reduce phase, the selector_spreading use `nodeNameToInfo[result[i].Host].Node()` directly，infact，the result[i].Host is null str；then the Node is nil；but  the scheduler in v1.11  do not deal with the error and return immediately in map phase, it deal with all errors after map & reduce all completed,  but I found this error deal in v1.19(new scheduling framwork)has been sovled, so thanks for reviewing this code


```
    
func (s *SelectorSpread) CalculateSpreadPriorityMap(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (schedulerapi.HostPriority, error) {
	var selectors []labels.Selector
	node := nodeInfo.Node()
	if node == nil {
		return schedulerapi.HostPriority{}, fmt.Errorf("node not found")
	}

	priorityMeta, ok := meta.(*priorityMetadata)
	if ok {
		selectors = priorityMeta.podSelectors
	} else {
		selectors = getSelectors(pod, s.serviceLister, s.controllerLister, s.replicaSetLister, s.statefulSetLister)
	}

	if len(selectors) == 0 {
		return schedulerapi.HostPriority{
			Host:  node.Name,
			Score: int(0),
		}, nil
	}

	count := int(0)
	for _, nodePod := range nodeInfo.Pods() {
		if pod.Namespace != nodePod.Namespace {
			continue
		}
		// When we are replacing a failed pod, we often see the previous
		// deleted version while scheduling the replacement.
		// Ignore the previous deleted version for spreading purposes
		// (it can still be considered for resource restrictions etc.)
		if nodePod.DeletionTimestamp != nil {
			glog.V(4).Infof("skipping pending-deleted pod: %s/%s", nodePod.Namespace, nodePod.Name)
			continue
		}
		matches := false
		for _, selector := range selectors {
			if selector.Matches(labels.Set(nodePod.ObjectMeta.Labels)) {
				matches = true
				break
			}
		}
		if matches {
			count++
		}
	}
	return schedulerapi.HostPriority{
		Host:  node.Name,
		Score: int(count),
	}, nil
}
```


```
	appendError := func(err error) {
		mu.Lock()
		defer mu.Unlock()
		errs = append(errs, err)
	}

	results := make([]schedulerapi.HostPriorityList, len(priorityConfigs), len(priorityConfigs))

	for i, priorityConfig := range priorityConfigs {
		if priorityConfig.Function != nil {
			// DEPRECATED
			wg.Add(1)
			go func(index int, config algorithm.PriorityConfig) {
				defer wg.Done()
				var err error
				results[index], err = config.Function(pod, nodeNameToInfo, nodes)
				if err != nil {
					appendError(err)
				}
			}(i, priorityConfig)
		} else {
			results[i] = make(schedulerapi.HostPriorityList, len(nodes))
		}
	}
	processNode := func(index int) {
		nodeInfo := nodeNameToInfo[nodes[index].Name]
		var err error
		for i := range priorityConfigs {
			if priorityConfigs[i].Function != nil {
				continue
			}
			start := time.Now()
			results[i][index], err = priorityConfigs[i].Map(pod, meta, nodeInfo)
			glog.V(4).Infof("PrioritizeNodes (%v, %v, %v, %v, %v) map func, score: %v,  started: %v, total time: %v", pod.Namespace, pod.Name, priorityConfigs[i].Name, priorityConfigs[i].Weight, nodes[index].Name, results[i][index].Score, start, time.Since(start))
			if err != nil {
				appendError(err)
				return
			}
		}
	}
	workqueue.Parallelize(16, len(nodes), processNode)

	for i, priorityConfig := range priorityConfigs {
		glog.V(4).Infof("PrioritizeNodes (%v, %v, %v, %v) hostPriorityList before reduce normalize: %v", pod.Namespace, pod.Name, priorityConfig.Name, priorityConfig.Weight, results[i])
		if priorityConfig.Reduce == nil {
			continue
		}
		wg.Add(1)
		go func(index int, config algorithm.PriorityConfig) {
			defer wg.Done()
			start := time.Now()
			if err := config.Reduce(pod, meta, nodeNameToInfo, results[index]); err != nil {
				appendError(err)
			}
			glog.V(4).Infof("PrioritizeNodes (%v, %v, %v, %v) reduce func started: %v, total time: %v", pod.Namespace, pod.Name, priorityConfigs[i].Name, priorityConfigs[i].Weight, start, time.Since(start))
			if glog.V(10) {
				for _, hostPriority := range results[index] {
					glog.Infof("%v -> %v: %v, Score: (%d)", pod.Name, hostPriority.Host, config.Name, hostPriority.Score)
				}
			}
			glog.V(4).Infof("PrioritizeNodes (%v, %v, %v, %v) hostPriorityList after reduce normalize: %v", pod.Namespace, pod.Name, config.Name, config.Weight, results[index])
		}(i, priorityConfig)
	}
	// Wait for all computations to be finished.
	wg.Wait()
	if len(errs) != 0 {
		return schedulerapi.HostPriorityList{}, errors.NewAggregate(errs)
	}

```
